File: get_cookies.js
const fs = require("fs");
const path = require("path");
const readline = require("readline");
const puppeteer = require("puppeteer");

const DIR = __dirname;
const COOKIE_FILE = path.join(DIR, "a.json");
const EDGE_PATH = "C:\\Program Files (x86)\\Microsoft\\Edge\\Application\\msedge.exe";
const DEBUG_DIR = path.join(DIR, "debug");

function getActiveEmail() {
  const idx = process.argv.indexOf("--email");
  if (idx >= 0 && process.argv[idx + 1]) {
    return process.argv[idx + 1];
  }

  const credsFile = path.join(DIR, "credentials.json");
  if (fs.existsSync(credsFile)) {
    try {
      const creds = JSON.parse(fs.readFileSync(credsFile, "utf8"));
      if (creds.email && creds.email !== "your_email@example.com") {
        return creds.email;
      }
    } catch (_) {}
  }

  return "default";
}

function getUserDataPath(email) {
  const safeName = email.replace(/[^a-zA-Z0-9_]/g, "_");
  return path.join(DIR, "user_data", safeName);
}

function loadAllCookies() {
  if (!fs.existsSync(COOKIE_FILE)) {
    return {};
  }

  try {
    const raw = fs.readFileSync(COOKIE_FILE, "utf8").trim();
    if (!raw) {
      return {};
    }
    const data = JSON.parse(raw);
    return Array.isArray(data) ? { [getActiveEmail()]: data } : data;
  } catch (err) {
    console.error(`Could not parse a.json: ${err.message}`);
    return {};
  }
}

function cookieKey(cookie) {
  return `${cookie.name || ""}|${cookie.domain || ""}|${cookie.path || "/"}`;
}

function isJioMartCookie(cookie) {
  return String(cookie.domain || "").includes("jiomart.com");
}

function normalizeCookies(cookieList) {
  const map = new Map();

  for (const cookie of cookieList || []) {
    if (!cookie || !cookie.name || !cookie.value) {
      continue;
    }

    const variants = [{ ...cookie }];
    if (
      isJioMartCookie(cookie) &&
      ["R.session", "app_location_details", "app_geolocation"].includes(cookie.name)
    ) {
      variants.push({
        ...cookie,
        domain: ".jiomart.com",
        path: "/",
      });
    }

    for (const variant of variants) {
      map.set(cookieKey(variant), variant);
    }
  }

  return Array.from(map.values());
}

function saveCookies(email, cookieList) {
  const allCookies = loadAllCookies();
  allCookies[email] = normalizeCookies(cookieList);
  fs.writeFileSync(COOKIE_FILE, JSON.stringify(allCookies, null, 2), "utf8");
}

function waitForEnter(prompt) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  return new Promise((resolve) => {
    rl.question(prompt, () => {
      rl.close();
      resolve();
    });
  });
}

function hasArg(name) {
  return process.argv.includes(name);
}

function ensureDebugDir() {
  if (!fs.existsSync(DEBUG_DIR)) {
    fs.mkdirSync(DEBUG_DIR, { recursive: true });
  }
}

async function applyBrowserPatches(page) {
  await page.evaluateOnNewDocument(() => {
    Object.defineProperty(navigator, "webdriver", { get: () => undefined });
    Object.defineProperty(navigator, "plugins", { get: () => [1, 2, 3, 4, 5] });
    Object.defineProperty(navigator, "languages", { get: () => ["en-US", "en"] });
    window.chrome = window.chrome || { runtime: {} };
  });
}

function attachDiagnostics(page, events) {
  page.on("console", (message) => {
    events.console.push({
      type: message.type(),
      text: message.text(),
    });
  });

  page.on("pageerror", (error) => {
    events.pageErrors.push(String(error.stack || error.message || error));
  });

  page.on("requestfailed", (request) => {
    events.requestFailures.push({
      url: request.url(),
      method: request.method(),
      failure: request.failure()?.errorText || "unknown",
      resourceType: request.resourceType(),
    });
  });

  page.on("response", (response) => {
    const status = response.status();
    if (status >= 400) {
      events.badResponses.push({
        url: response.url(),
        status,
        resourceType: response.request().resourceType(),
      });
    }
  });
}

async function getAllCookies(page) {
  const client = await page.target().createCDPSession();
  const result = await client.send("Network.getAllCookies");
  await client.detach();
  return result.cookies || [];
}

async function clickSignInIfVisible(page) {
  const selectors = [
    'a[href="/profile"]',
    'a[href*="/profile"]',
    "a[href*='login']",
    "#sign-in",
    ".sign-in",
    ".login-icon-content",
    "a.logged-user-name",
  ];

  for (const selector of selectors) {
    const handle = await page.$(selector);
    if (!handle) {
      continue;
    }
    try {
      await handle.click();
      console.log("Clicked Sign In/login element.");
      return true;
    } catch (_) {}
  }

  const clickedByText = await page.evaluate(() => {
    const nodes = Array.from(document.querySelectorAll("a, button, div, span"));
    const node = nodes.find((el) => /sign\s*in/i.test(el.textContent || ""));
    if (node) {
      node.click();
      return true;
    }
    return false;
  });

  if (clickedByText) {
    console.log("Clicked Sign In by text.");
  }
  return clickedByText;
}

async function runDebug(browser, page) {
  ensureDebugDir();

  const events = {
    console: [],
    pageErrors: [],
    requestFailures: [],
    badResponses: [],
  };
  attachDiagnostics(page, events);

  console.log("DEBUG: opening JioMart homepage...");
  let navigationError = null;
  try {
    await page.goto("https://www.jiomart.com/", {
      waitUntil: "domcontentloaded",
      timeout: 60000,
    });
  } catch (err) {
    navigationError = err.message;
    console.log(`DEBUG: homepage load error: ${err.message}`);
  }

  await new Promise((resolve) => setTimeout(resolve, 8000));

  console.log(`DEBUG: current URL after homepage load: ${page.url()}`);
  const clicked = await clickSignInIfVisible(page);
  console.log(`DEBUG: clicked sign-in: ${clicked ? "yes" : "no"}`);

  await new Promise((resolve) => setTimeout(resolve, 12000));

  const cookiesNow = await getAllCookies(page);
  const html = await page.content().catch((err) => `HTML capture failed: ${err.message}`);
  const title = await page.title().catch(() => "");
  const bodyText = await page.evaluate(() => document.body?.innerText || "").catch(() => "");
  const screenshotPath = path.join(DEBUG_DIR, "auth-debug.png");
  const htmlPath = path.join(DEBUG_DIR, "auth-debug.html");
  const jsonPath = path.join(DEBUG_DIR, "auth-debug.json");

  await page.screenshot({ path: screenshotPath, fullPage: true }).catch((err) => {
    events.pageErrors.push(`screenshot failed: ${err.message}`);
  });
  fs.writeFileSync(htmlPath, html, "utf8");
  fs.writeFileSync(
    jsonPath,
    JSON.stringify(
      {
        navigationError,
        finalUrl: page.url(),
        title,
        bodyTextLength: bodyText.length,
        bodyTextPreview: bodyText.slice(0, 500),
        htmlLength: html.length,
        cookieNames: cookiesNow.map((cookie) => ({
          name: cookie.name,
          domain: cookie.domain,
          path: cookie.path,
          expires: cookie.expires,
        })),
        events,
      },
      null,
      2,
    ),
    "utf8",
  );

  console.log(`DEBUG: saved screenshot: ${screenshotPath}`);
  console.log(`DEBUG: saved html: ${htmlPath}`);
  console.log(`DEBUG: saved report: ${jsonPath}`);
  console.log(`DEBUG: final URL: ${page.url()}`);
  console.log(`DEBUG: title: ${title}`);
  console.log(`DEBUG: body text length: ${bodyText.length}`);
  console.log(`DEBUG: page errors: ${events.pageErrors.length}`);
  console.log(`DEBUG: failed requests: ${events.requestFailures.length}`);
  console.log(`DEBUG: bad responses: ${events.badResponses.length}`);

  await browser.close();
}

async function main() {
  const email = getActiveEmail();
  const userDataDir = getUserDataPath(email);

  console.log(`Starting Puppeteer cookie capture for ${email}...`);
  console.log(`Using persistent profile directory: ${userDataDir}`);

  if (!fs.existsSync(EDGE_PATH)) {
    throw new Error(`Edge executable not found: ${EDGE_PATH}`);
  }

  const browser = await puppeteer.launch({
    executablePath: EDGE_PATH,
    userDataDir,
    headless: false,
    defaultViewport: null,
    ignoreDefaultArgs: ["--enable-automation", "--no-sandbox"],
    args: [
      "--start-maximized",
      "--disable-blink-features=AutomationControlled",
      "--no-first-run",
      "--no-default-browser-check",
    ],
  });

  const pages = await browser.pages();
  const page = pages[0] || (await browser.newPage());
  await applyBrowserPatches(page);

  if (hasArg("--debug")) {
    await runDebug(browser, page);
    return;
  }

  console.log("Opening JioMart homepage...");
  try {
    await page.goto("https://www.jiomart.com/", {
      waitUntil: "domcontentloaded",
      timeout: 60000,
    });
  } catch (err) {
    console.log(`Initial page load did not finish cleanly: ${err.message}`);
  }

  await new Promise((resolve) => setTimeout(resolve, 5000));
  const beforeCookies = await getAllCookies(page);
  const hasSessionBefore = beforeCookies.some((cookie) => cookie.name === "R.session");

  if (hasSessionBefore) {
    console.log("R.session already exists in this browser profile.");
  } else {
    console.log("R.session not found yet. Opening login manually if possible...");
    await clickSignInIfVisible(page);
  }

  console.log("\n============================================================");
  console.log("INSTRUCTIONS:");
  console.log("1. If page is blank, type https://www.jiomart.com/ in this browser and press Enter.");
  console.log("2. Sign in and complete OTP in this same automated Edge window.");
  console.log("3. After login is complete, return here and press ENTER.");
  console.log("============================================================\n");

  await waitForEnter("Press Enter here AFTER successful login...");

  const savedCookies = await getAllCookies(page);
  saveCookies(email, savedCookies);

  const sessionCookie = savedCookies.find((cookie) => cookie.name === "R.session");
  console.log(`Saved ${savedCookies.length} cookies to a.json for ${email}.`);
  console.log(`R.session present: ${sessionCookie ? "yes" : "no"}`);
  if (sessionCookie) {
    console.log(`R.session domain: ${sessionCookie.domain || "(none)"}`);
    console.log(`R.session path: ${sessionCookie.path || "/"}`);
    console.log(`R.session expires: ${sessionCookie.expires || "session/server-controlled"}`);
  }

  await browser.close();
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: get_cookies.py
import json
import os
import re
import sys

if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")
if hasattr(sys.stderr, "reconfigure"):
    sys.stderr.reconfigure(encoding="utf-8")

from playwright.sync_api import sync_playwright

import cookies


DIR = os.path.dirname(os.path.abspath(__file__))
DEFAULT_BROWSER_CHANNEL = os.environ.get("JIOMART_BROWSER_CHANNEL", "msedge")
BROWSER_ARGS = [
    "--start-maximized",
    "--disable-blink-features=AutomationControlled",
    "--disable-dev-shm-usage",
    "--no-first-run",
    "--no-default-browser-check",
]


def get_user_data_path(email):
    safe_name = re.sub(r"[^a-zA-Z0-9_]", "_", email)
    return os.path.join(DIR, "user_data", safe_name)


def launch_context(playwright, user_data_dir, headless=False):
    launch_options = {
        "user_data_dir": user_data_dir,
        "headless": headless,
        "no_viewport": True,
        "args": BROWSER_ARGS,
        "ignore_default_args": ["--enable-automation", "--no-sandbox"],
    }

    if DEFAULT_BROWSER_CHANNEL:
        try:
            print(f"Launching browser with channel: {DEFAULT_BROWSER_CHANNEL}")
            return playwright.chromium.launch_persistent_context(
                channel=DEFAULT_BROWSER_CHANNEL,
                **launch_options,
            )
        except Exception as e:
            print(f"Could not launch {DEFAULT_BROWSER_CHANNEL}; falling back to bundled Chromium.")
            print(f"Launch error: {e}")

    print("Launching bundled Playwright Chromium.")
    return playwright.chromium.launch_persistent_context(**launch_options)


def apply_browser_patches(context):
    context.add_init_script(
        """
        Object.defineProperty(navigator, 'webdriver', {
          get: () => undefined
        });
        Object.defineProperty(navigator, 'plugins', {
          get: () => [1, 2, 3, 4, 5]
        });
        Object.defineProperty(navigator, 'languages', {
          get: () => ['en-US', 'en']
        });
        window.chrome = window.chrome || { runtime: {} };
        """
    )


def save_context_cookies(context, email):
    playwright_cookies = context.cookies()
    cookies.save_cookies(email, playwright_cookies)
    has_session = any(c.get("name") == "R.session" for c in playwright_cookies)
    print(f"\nSaved {len(playwright_cookies)} cookies to a.json for {email}.")
    print(f"R.session present: {'yes' if has_session else 'no'}")
    return playwright_cookies


def close_location_popup(page):
    print("Checking for location popup...")
    try:
        page.wait_for_timeout(3000)
        locators = [
            page.locator("button:has-text('Select Location Manually')"),
            page.locator("div:has-text('Select Location Manually')").last,
            page.locator("button.close"),
            page.locator("[aria-label='Close']"),
            page.locator(".modal-close"),
            page.locator(".jmart-modal-close"),
            page.locator("div.modal-header button"),
        ]
        for loc in locators:
            try:
                if loc.count() > 0 and loc.first.is_visible():
                    loc.first.click()
                    print("Location popup closed successfully.")
                    break
            except Exception:
                continue
    except Exception as e:
        print(f"No location popup or failed to close it automatically: {e}")


def is_logged_in(context, page):
    print("Checking if user is already logged in...")
    page.wait_for_timeout(3000)

    playwright_cookies = context.cookies()
    has_session_cookie = any(c.get("name") == "R.session" for c in playwright_cookies)

    try:
        sign_in_link = page.locator("a:has-text('Sign In'), #sign-in, .sign-in").first
        sign_in_visible = sign_in_link.count() > 0 and sign_in_link.is_visible()
    except Exception:
        sign_in_visible = False

    return has_session_cookie and not sign_in_visible


def open_login_if_needed(page):
    print("Opening login drawer if needed...")
    sign_in_locators = [
        page.locator("a:has-text('Sign In')"),
        page.locator("button:has-text('Sign In')"),
        page.locator("#sign-in"),
        page.locator(".sign-in"),
        page.locator("a[href*='login']"),
        page.locator(".login-icon-content"),
        page.locator("a.logged-user-name"),
    ]

    for loc in sign_in_locators:
        try:
            if loc.count() > 0 and loc.first.is_visible():
                loc.first.click()
                print("Sign In drawer/modal opened.")
                return
        except Exception:
            continue

    print("Could not click Sign In automatically. Please click it manually in the browser.")


def main():
    email = cookies.get_active_email()
    print(f"Starting JioMart login process for {email}...")

    user_data_dir = get_user_data_path(email)
    print(f"Using persistent profile directory: {user_data_dir}")

    with sync_playwright() as p:
        context = launch_context(p, user_data_dir, headless=False)
        apply_browser_patches(context)
        page = context.pages[0] if context.pages else context.new_page()

        print("Opening JioMart homepage...")
        try:
            page.goto("https://www.jiomart.com/", wait_until="domcontentloaded", timeout=60000)
        except Exception as e:
            print(f"Initial page load did not finish cleanly: {e}")
            print("If the page is visible in the browser, continue manually and then press ENTER here.")

        page.wait_for_timeout(5000)
        close_location_popup(page)

        if is_logged_in(context, page):
            print("\nDetected active login session in browser profile.")
        else:
            print("\nUser is not logged in. Manual login is needed.")
            open_login_if_needed(page)

            print("\n" + "=" * 60)
            print("INSTRUCTIONS:")
            print("1. If the page is stuck, press Ctrl+R in the opened browser once.")
            print("2. Click Sign In if the login drawer is not open.")
            print("3. Enter mobile number and verify OTP in this browser.")
            print("4. After successful login, come back here and press ENTER.")
            print("=" * 60 + "\n")

            input("Press Enter here AFTER you have successfully logged in in the browser...")

        save_context_cookies(context, email)
        context.close()


if __name__ == "__main__":
    main()
